core-client/src/transports/duplex.rs

```diff
@@ -104,20 +142,110 @@ where
 				Ok(Async::NotReady) => break,
 				Err(err) => Err(err)?,
 			};
+			log::debug!("incoming: {}", response_str);
+			for response in super::parse_response(&response_str)? {
+				self.incoming.push_back(response);
+			}
+		}
+
+		// Handle incoming queue.
+		log::debug!("handle incoming");
+		loop {
+			match self.incoming.pop_front() {
+				Some((id, result, method)) => match method {
+					// is a notification
+					Some(method) => {
+						match self.subscriptions.get_mut(&method) {
+							Some(subscription) => match subscription.channel.poll_ready() {
+								Ok(Async::Ready(())) => {
+									subscription.channel.try_send(result).expect("shouldn't fail; qed");
```

Comment:
the error type is (), I really don't think it can fail

Reply:
I'm not discussing if that fails or not, but the rule is that `expect` contains a proof to why it doesn't fail, something like:
```
.expect("`the error type is (); mpsc::Sender::try_send never fails according to the docs; qed")
```
is way better than `shouldn't fail`, cause it states your assumptions.

That said, I don't think it actually never fails (and the error type is `()`):
https://docs.rs/futures/0.1.27/futures/sync/mpsc/struct.Sender.html#method.try_send

There at least two cases where it will fail:
1. The channel is full (i.e. it would block)
2. The channel is disconnected (i.e. the receiving end was dropped)

If you want to keep the `expect` you need to provide a proof that none of the two conditions will ever occur.

Reply:
Ok I think the assumption I made was that poll_ready returns Ok. I updated the expect message to reflect this, let me know if that works.